velox/type/DecimalUtils.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "DecimalUtils.h"
+
+namespace facebook::velox {
+
+static const facebook::velox::int128_t POWERS_OF_TEN[]{
+    1,
+    10,
+    100,
+    1000,
+    10000,
+    100000,
+    1000000,
+    10000000,
+    100000000,
+    1000000000,
+    10000000000,
+    100000000000,
+    1000000000000,
+    10000000000000,
+    100000000000000,
+    1000000000000000,
+    10000000000000000,
+    100000000000000000,
+    1000000000000000000,
+    1000000000000000000 * (int128_t)10,
+    1000000000000000000 * (int128_t)100,
+    1000000000000000000 * (int128_t)1000,
+    1000000000000000000 * (int128_t)10000,
+    1000000000000000000 * (int128_t)100000,
+    1000000000000000000 * (int128_t)1000000,
+    1000000000000000000 * (int128_t)10000000,
+    1000000000000000000 * (int128_t)100000000,
+    1000000000000000000 * (int128_t)1000000000,
+    1000000000000000000 * (int128_t)10000000000,
+    1000000000000000000 * (int128_t)100000000000,
+    1000000000000000000 * (int128_t)1000000000000,
+    1000000000000000000 * (int128_t)10000000000000,
+    1000000000000000000 * (int128_t)100000000000000,
+    1000000000000000000 * (int128_t)1000000000000000,
+    1000000000000000000 * (int128_t)10000000000000000,
+    1000000000000000000 * (int128_t)100000000000000000,
+    1000000000000000000 * (int128_t)1000000000000000000,
+    1000000000000000000 * (int128_t)1000000000000000000 * (int128_t)10,
+    1000000000000000000 * (int128_t)1000000000000000000 * (int128_t)100};
+
+namespace {
+std::string formatDecimal(
+    uint8_t scale,
+    facebook::velox::int128_t unscaledValue) {
+  VELOX_CHECK_LE(0, scale);
```

Comment:
If this function to be used in cast operation, these checks will be expensive.

Reply:
Agree, we have these checks at the of Decimal Type creation. It is less likely to cross the max precision limits again. Will remove these checks for now.

Reply:
Make these debug checks: `VELOX_DCHECK_LE`